### DISCUSSION TITLE
How do I copy files after bundling for each entry?

### DISCUSSION THREAD

**Discussion by vicary** [floydspace/serverless-esbuild#214] [Q&A]
So this is a bit tricky, the need comes from dynamic imports from `vm2` which is one of the many problematic but commonly used dependencies. Some other examples would be `coffee-script` and `formiable`.

For more context about the problem please see https://github.com/evanw/esbuild/issues/495, it essentially boils down to a requirement of copying necessary files ourselves after bundling. But at the same time I don't want to blindly copy to all entries (I have 300 lambdas), but only copy when the resolution hits a certain path. i.e. Copy `contextify.js` and `sandbox.js` when `vm2` is hit at least once for the current entry.

Does `serverless-esbuild` support such a per entry build hook right now? If not, is it possible to add one?

## Alternatives considered

1. esbuild may enable extra APIs for plugins to do that in the future but the discussion is progressing really slow https://github.com/evanw/esbuild/issues/700
2. Nelify as [a fork](https://github.com/netlify/esbuild/blob/netlify/NETLIFY.md) that adds `onDynamicImport` but it require quite a lot of hacks for a sub-dependency to know which entry it's currently bundling, also `serverless-esbuild` needs to allow a custom impl of esbuild.
3. Copying artifacts by enabling `metafile` with an `onEnd(...)` callback as a custom plugin, makes the packaging step including all copies from other entries. Say I have 300 lambdas, every entry will contain the copied assets of the other 299 lambas. This instantly explodes the resulting bundle from GB levels into TB levels.

# Update

Turns out I don't actually need to copy those files because I don't need `vm2` at all. The exact tree was `auth0` -> `rest-facade` -> `superagent-proxy` -> `proxy-agent` -> `pac-proxy-agent` -> `pac-resolver` -> `degenerator` -> `vm2`.

Both `rest-facade` and `degenerator` already updated to make the downstream requirement optional, see https://github.com/auth0/node-auth0/issues/657#issuecomment-953120110 for the workaround.

But this question remains valid for those who actually need HTTP proxy, answers are still welcomed and I'll mark it for others if it works.

**Comment by samchungy** [comment]
There is an option to bundle functions individually like this example [here](https://github.com/floydspace/serverless-esbuild/blob/master/examples/individually/serverless.yml)

By entry do you mean per function handler path? 

If so, you can combine it with function package [patterns](https://www.serverless.com/framework/docs/providers/aws/guide/packaging#packaging-functions-separately) to package those extra files you need on a per function basis?

**Reply by vicary** [reply]
We are already using `package.individually`, and by the look of `onEnd({ metafile })` it is already generating one entry per Lambda so it is working fine.

What I want is that for each entry, I would like to copy a few more files where `vm2` tries to dynamically require in runtime (namely, `contextify.js` and `sandbox.js`). The intermediate structure is one directory per Lambda, and only one js file is there before archiving into ZIP and upload to S3 for deployment.

`vm2` is importing those files like they are in the same directory `node_modules/vm2/lib`, once it was bundled it will try to import file the lambda root, e.g. `myFunction/handler/index.js` will require `contextify.js` from `myFunction/handler`. If I am understanding it right, [patterns](https://www.serverless.com/framework/docs/providers/aws/guide/packaging#packaging-functions-separately) will directly include matched files keeping the same directory structure, i.e. It stays in `node_modules/vm2/lib`, where my bundle still can't locate.

The closest thing I got to is alternative 3, where I copy the files into the same directory for each entry output. But it seems that the packaging step only excludes entry files (the other 299 lambda handler js) and includes all remaining files (my 600 copied files) into the final archive, for each entry. Such that I will have 300 ZIP all with 601 files inside, this step alone takes the disk requirements into TB level.
